core/src/main/java/kafka/server/NetworkPartitionMetadataClient.java

```diff
@@ -0,0 +1,287 @@
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.message.ListOffsetsRequestData.ListOffsetsPartition;
+import org.apache.kafka.common.message.ListOffsetsRequestData.ListOffsetsTopic;
+import org.apache.kafka.common.message.ListOffsetsResponseData.ListOffsetsPartitionResponse;
+import org.apache.kafka.common.message.ListOffsetsResponseData.ListOffsetsTopicResponse;
+import org.apache.kafka.common.network.ListenerName;
+import org.apache.kafka.common.protocol.Errors;
+import org.apache.kafka.common.requests.ListOffsetsRequest;
+import org.apache.kafka.common.requests.ListOffsetsResponse;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.coordinator.group.PartitionMetadataClient;
+import org.apache.kafka.metadata.MetadataCache;
+import org.apache.kafka.server.util.InterBrokerSendThread;
+import org.apache.kafka.server.util.RequestAndCompletionHandler;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.stream.Collectors;
+
+public class NetworkPartitionMetadataClient implements PartitionMetadataClient {
+
+    private static final Logger log = LoggerFactory.getLogger(NetworkPartitionMetadataClient.class);
+
+    private final MetadataCache metadataCache;
+    private final SendThread sendThread;
+    private final Time time;
+    private final ListenerName listenerName;
+
+    public NetworkPartitionMetadataClient(
+        MetadataCache metadataCache,
+        KafkaClient networkClient,
+        Time time,
+        ListenerName listenerName
+    ) {
+        this.metadataCache = metadataCache;
+        this.time = time;
+        this.listenerName = listenerName;
+        this.sendThread = new SendThread(
+            "NetworkPartitionMetadataClientSendThread",
+            networkClient,
+            Math.toIntExact(CommonClientConfigs.DEFAULT_SOCKET_CONNECTION_SETUP_TIMEOUT_MAX_MS),  //30 seconds
+            this.time
+        );
+        this.sendThread.start();
+    }
+
+    @Override
+    public Map<TopicPartition, CompletableFuture<OffsetResponse>> listLatestOffsets(
+        Set<TopicPartition> topicPartitions
+    ) {
+        if (topicPartitions == null || topicPartitions.isEmpty()) {
+            return Map.of();
+        }
+
+        // Map to store futures for each TopicPartition
+        Map<TopicPartition, CompletableFuture<OffsetResponse>> futures = new HashMap<>();
+
+        // Group TopicPartitions by leader node
+        Map<Node, List<TopicPartition>> partitionsByNode = new HashMap<>();
+
+        for (TopicPartition tp : topicPartitions) {
+            CompletableFuture<OffsetResponse> future = new CompletableFuture<>();
+            futures.put(tp, future);
+
```

Comment:
Do not add the future yet, it's not helpful to read seeing multiple passes in different maps.
```suggestion
```